osu.Game/Beatmaps/BeatmapDifficultyManager.cs

```diff
@@ -114,6 +115,26 @@ public StarDifficulty GetDifficulty([NotNull] BeatmapInfo beatmapInfo, [CanBeNul
             return computeDifficulty(key, beatmapInfo, rulesetInfo);
         }
 
+        /// <summary>
+        /// Calculates performance for the given <see cref="ScoreInfo"/> on a given <see cref="WorkingBeatmap"/>.
+        /// </summary>
+        /// <param name="beatmap">The <see cref="WorkingBeatmap"/> to do the calculation on. </param>
+        /// <param name="score">The score to do the calculation on. </param>
+        /// <param name="token">An optional <see cref="CancellationToken"/> to cancel the operation.</param>
+        public async Task<double> CalculatePerformance([NotNull] WorkingBeatmap beatmap, [NotNull] ScoreInfo score, CancellationToken token = default)
+        {
+            return await Task.Factory.StartNew(() =>
+            {
+                if (token.IsCancellationRequested)
+                    return default;
+
+                var calculator = score.Ruleset.CreateInstance().CreatePerformanceCalculator(beatmap, score);
```

Comment:
In addition to the above caching, this should use the result of `GetDifficultyAsync()` too otherwise it's recomputing difficulty every time, even though we're pretty much guaranteed to already have a cached version. Maybe `StarDifficulty` should contain the full difficulty attributes.